Performance-Tracing.md

```diff
@@ -0,0 +1,156 @@
+    If none of the suggestions help, please file an [issue](https://github.com/microsoft/TypeScript/issues), including code that reproduces the slowdown you're seeing and instructions for compiling it.
+
+
+## Advanced
+
+At this point, it's going to be pretty hard to keep going if you don't have a good mental model of how types are represented in the compiler.
+On top of that, the structure of this output is very likely to change in TS 4.2.
+You've been warned...
+
+To dig further on your own, you'll need to start looking at types.
+
+![Type checking call stack](https://raw.githubusercontent.com/wiki/Microsoft/TypeScript/images/tracingTypes.png)
+
+Here's a slow section with no `checkExpression` boxes in the (partial) stack.
+Instead, it's gone straight into type checking.
+To work out which code fragment triggered this work and what it's doing, you'll need to open `types.json`.
+From the metadata, we can see that the types of interest have IDs `20441` and `20440`.
+Here are the corresponding entries from `types.json`.
+(At the time of writing, the type with ID `N` is on line `N`.)
+
+```
+{"id":20440,"symbolName":"NamedExoticComponent","recursionId":30,"instantiatedType":146,"typeArguments":[20437],"firstDeclaration":{"path":"PROJECT_ROOT/node_modules/@types/react/index.d.ts","start":{"line":359,"character":6},"end":{"line":363,"character":6}},"flags":["524288"]},
+{"id":20441,"symbolName":"NamedExoticComponent","recursionId":30,"instantiatedType":146,"typeArguments":[20434],"firstDeclaration":{"path":"PROJECT_ROOT/node_modules/@types/react/index.d.ts","start":{"line":359,"character":6},"end":{"line":363,"character":6}},"flags":["524288"]},
+```
+
+This data gives us some sense of what they look like and a location where we can review the declaration.
+In this case, we're comparing two different instantiations of [NamedExoticComponent&lt;P&gt;](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/599ef3c6e2142d87b6b4bcf030fb798504c3a468/types/react/index.d.ts#L361) - one instantiated with type `20434` and the other with `20437`.
+These types are also described in `types.json`, so keep looking up more types recursively until you understand what's going on.
+From the file path in `checkSourceFile` and the types we've just pulled out of `types.json`, we can probably guess approximately which code fragment was involved (knowing in advance that it's neither an expression, nor a variable declaration, which would have gotten a box).
+
+Look hard enough and you'll be able to create a toy program that reproduces the performance issue you're seeing in your real code.
+We'd love to hear about it but, if you still can't share, the next step is to play around with alternate declarations and `any` type assertions until you've sped up the toy program.
+Then you can apply the same fix or mitigation to your real code.
+
+*Aside:* the [edge://tracing](edge://tracing) tracing controls are rather clumsy.
```

Comment:
Maybe mention in the above something like "see the bottom for using the usual devtools profiler"?

Also, maybe add that we'll try to make it work as the main tool used for visualization, if possible?

Reply:
I was de-emphasizing it because I thought you wanted to use a different tool.  If you've used it for an investigation and feel that it's up to the task, I don't mind moving this up.